### ISSUE TITLE
Incorrect type inference in loops

### ISSUE THREAD

**Issue by Molkree** [microsoft/pylance-release#2552]
## Environment data

-   Language Server version: v2022.3.4 - v2022.4.0
-   OS and version: Win10
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.10

## Expected behaviour

No error.

## Actual behaviour

`Type of "high" is partially unknown`
`  Type of "high" is "int | Unknown"`

## Code Snippet / Additional information

```python
def find_min(nums: list[int]) -> int:
    low = 0
    high = len(nums) - 1
    while low < high:
        mid = (low + high) // 2
        if nums[mid] > nums[high]:
            low = mid + 1
        elif nums[mid] < nums[high]:
            high = mid
        else:
            high -= 1  # <- error here
    return nums[low]
```
This might have the same root issue as the #2549 as it is also in a loop but this issue is present since v2020.3.4 and another one only since v2020.4.0 so I have opted to open a new ticket.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release. I've added a new unit test based on your code sample to help prevent future regression.

**Comment by heejaechang** [comment]
This issue has been fixed in version 2022.4.1, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202241-13-April-2022)
